csrc/ops/indexing.cpp

```diff
@@ -0,0 +1,278 @@
+
+  // The shape of output tensor is same as self tensor.
+  std::vector<IterDomain*> out_domain;
+  for (const auto i : c10::irange(self_dom.size())) {
+    out_domain.push_back(
+        IterDomainBuilder(self_dom[i])
+            .iter_type(
+                self_dom[i]->getIterType() == IterType::Iteration
+                    ? IterType::GatherScatter
+                    : self_dom[i]->getIterType())
+            .build());
+  }
+
+  TensorView* out_tensor = IrBuilder::create<TensorView>(
+      IrBuilder::create<TensorDomain>(
+          out_domain, TensorDomain::getContiguityFilledWith(out_domain, true)),
+      self->getDataType().value());
+
+  IrBuilder::create<ScatterOp>(
+      type, out_tensor, self, dim, index, src, out_domain[dim]);
+  return out_tensor->as<TensorView>();
+}
+
+TensorView* scatter(
+    TensorView* self,
+    int dim,
+    TensorView* index,
+    TensorView* src) {
+  return scatterOp(ScatterOpType::Set, self, dim, index, src);
+}
+
+TensorView* take_along_axis(TensorView* inp, TensorView* index, int64_t dim) {
+  const auto inp_domain =
+      TensorDomain::noReductions(inp->getMaybeRFactorDomain());
+  const auto idx_domain =
+      TensorDomain::noReductions(index->getMaybeRFactorDomain());
+
+  TORCH_CHECK(
+      !inp_domain.empty(), "take_along_axis can not be applied to 0d tensor.");
+  TORCH_CHECK(
+      idx_domain.size() == inp_domain.size(),
+      "The input and index tensor must have the same dimensions for take_along_axis");
+
+  if (dim < 0) {
+    dim += (int)idx_domain.size();
+  }
+
+  TORCH_CHECK(
+      dim >= 0 && dim < (int)inp_domain.size(),
+      "take_along_axis on invalid axis, received: ",
+      dim,
+      " however tensor view only has ",
+      inp_domain.size(),
+      " non-reduction dims.");
+
+  std::vector<IterDomain*> out_domain(idx_domain.size());
+
+  for (const auto i : c10::irange(idx_domain.size())) {
+    auto inp_id = inp_domain.at(i);
+    auto idx_id = idx_domain.at(i);
+
+    TORCH_CHECK(
+        !inp_id->maybePartial(),
+        "Partial domain not supported: ",
+        inp_id->toString());
+    TORCH_CHECK(
+        !idx_id->maybePartial(),
+        "Partial domain not supported: ",
+        idx_id->toString());
+
+    TORCH_CHECK(
+        inp_id->getIterType() != IterType::Iteration ||
+            inp_id->getIterType() != IterType::Broadcast,
+        "Unsupported IterType of an input domian: ",
+        inp_id->toString());
+    TORCH_CHECK(
+        idx_id->getIterType() != IterType::Iteration ||
+            idx_id->getIterType() != IterType::Broadcast,
+        "Unsupported IterType of an index domian: ",
+        idx_id->toString());
+
+    // When the input ID is a broadcat with no expanded extent, the
+    // extent as well as the type of the output ID is determined by the index ID
+    if (static_cast<int>(i) == dim ||
+        (inp_id->isBroadcast() && !inp_id->hasExpandedExtent())) {
+      out_domain.at(i) = IterDomainBuilder(idx_id).build();
```

Comment:
If `inp_id` is broadcast and `idx_id` is concrete, then the output should be concrete, regardless of if `inp_id` has expand or not.

Reply:
Updated. Does this look good? [c4b0341](https://github.com/NVIDIA/Fuser/pull/206/commits/c4b0341ca6ae1f05c0ef83d7331b0dae1ec356fc)